gusto/time_discretisation.py

```diff
@@ -1084,3 +1090,463 @@ def apply(self, x_out, *x_in):
+
+    The AB2 timestepping method for operator F is written as
+    y^(n+1) = y^n + dt*((3/2)*F[y^(n) - (1/2)*F[y^(n-1)])
+    """
+    @property
+    def nlevels(self):
+        return 2
+
+    @property
+    def rhs0(self):
+        """Set up the discretisation's right hand side."""
+        r = self.residual.label_map(
+            all_terms,
+            map_if_true=replace_subject(self.x1, self.idx))
+        r = r.label_map(lambda t: t.has_label(time_derivative),
+                        map_if_false=lambda t: -self.dt*t)
+
+        return r.form
+
+    @property
+    def lhs(self):
+        """Set up the discretisation's left hand side (the time derivative)."""
+        return super(AB2, self).lhs
+
+    @property
+    def rhs(self):
+        """Set up the discretisation's right hand side."""
+        rn = self.residual.label_map(all_terms,
+                                     map_if_true=replace_subject(self.x1, self.idx))
+        rn = rn.label_map(lambda t: t.has_label(time_derivative),
+                          map_if_false=lambda t: -(3/2.)*self.dt*t)
+        rnm1 = self.residual.label_map(lambda t: t.has_label(time_derivative),
+                                       map_if_true=drop,
+                                       map_if_false=replace_subject(self.xnm1, self.idx))
+        rnm1 = rnm1.label_map(all_terms,
+                              map_if_true=lambda t: (1/2.)*self.dt*t)
+        r = rn + rnm1
+        return r.form
+
+    @property
+    def solver0(self):
+        """Set up the problem and the solver."""
+        # setup solver using lhs and rhs defined in derived class
+        problem = NonlinearVariationalProblem(self.lhs-self.rhs0, self.x_out, bcs=self.bcs)
+        solver_name = self.field_name+self.__class__.__name__+"0"
+        return NonlinearVariationalSolver(problem, solver_parameters=self.solver_parameters, options_prefix=solver_name)
+
+    @property
+    def solver(self):
+        """Set up the problem and the solver."""
+        # setup solver using lhs and rhs defined in derived class
+        problem = NonlinearVariationalProblem(self.lhs-self.rhs, self.x_out, bcs=self.bcs)
+        solver_name = self.field_name+self.__class__.__name__
+        return NonlinearVariationalSolver(problem, solver_parameters=self.solver_parameters, options_prefix=solver_name)
+
+    def apply(self, x_out, *x_in):
+        """
+        Apply the time discretisation to advance one whole time step.
+
+        Args:
+            x_out (:class:`Function`): the output field to be computed.
+            x_in (:class:`Function`): the input field(s).
+        """
+        if self.initial_timesteps < self.nlevels-1:
+            self.initial_timesteps += 1
+            solver = self.solver0
+        else:
+            solver = self.solver
+        self.xnm1.assign(x_in[0])
+        self.x1.assign(x_in[1])
+        solver.solve()
+        x_out.assign(self.x_out)
+
+
+class AB3(MultilevelTimeDiscretisation):
```

Comment:
Is there a general way of writing AB2 and AB3? Is it best to implement them as different classes or could they be the same class with an optional argument to indicate the number of stages?

Reply:
I've created two more general time discretisations AdamsMoulton and AdamsBashforth that take in order as an argument, and run the corresponding method. We now have up to AB5 and AM4